addons/hr_holidays/models/hr_leave_allocation.py

```diff
@@ -84,18 +93,95 @@ def _default_holiday_status_id(self):
+    interval_unit = fields.Selection([
+        ('weeks', 'Week(s)'),
+        ('months', 'Month(s)'),
+        ('years', 'Year(s)')
+        ], string="Unit of time between two intervals", default='weeks', readonly=True, states={'draft': [('readonly', False)], 'confirm': [('readonly', False)]})
+    nextcall = fields.Date("Date of the next accrual allocation", default=False)
 
     _sql_constraints = [
         ('type_value', "CHECK( (holiday_type='employee' AND employee_id IS NOT NULL) or (holiday_type='category' AND category_id IS NOT NULL) or (holiday_type='department' AND department_id IS NOT NULL) )",
          "The employee, department or employee category of this request is missing. Please make sure that your user login is linked to an employee."),
         ('date_check', "CHECK ( number_of_days_temp >= 0 )", "The number of days must be greater than 0."),
     ]
 
+    @api.model
+    def _update_accrual(self):
+        """
+            Method called by the cron task in order to increment the number_of_days when
+            necessary.
+        """
+        def to_tz(d, tz_name):
+            tz = pytz.timezone(tz_name) if tz_name else pytz.UTC
+            return pytz.UTC.localize(d.replace(tzinfo=None), is_dst=False).astimezone(tz).replace(tzinfo=None)
+
+        today = to_tz(fields.Datetime.from_string(fields.Datetime.now()), self.env.user.tz).date()
+
+        holidays = self.search([('accrual', '=', True), ('state', '=', 'validate'),
+                                '|', ('date_to', '=', False), ('date_to', '>', fields.Datetime.now()),
+                                '|', ('nextcall', '=', False), ('nextcall', '<=', today)])
+
+        for holiday in holidays:
+            values = {}
+
+            leaves = holiday.number_per_interval
+            if holiday.unit_per_interval == 'hours':
+                leaves = holiday.number_per_interval / holiday.employee_id.resource_calendar_id.hours_per_day
+
+            delta = relativedelta(days=0)
+
+            if holiday.interval_unit == 'weeks':
+                delta = relativedelta(weeks=holiday.interval_number)
+            if holiday.interval_unit == 'months':
+                delta = relativedelta(months=holiday.interval_number)
+            if holiday.interval_unit == 'years':
+                delta = relativedelta(years=holiday.interval_number)
+
+            values['nextcall'] = (holiday.nextcall if holiday.nextcall else today) + delta
+
+            period_start = datetime.combine(today, time(0, 0, 0)) - delta
+            period_end = datetime.combine(today, time(0, 0, 0))
+
+            # We have to check when the employee has been created
+            # in order to not allocate him/her too much leaves
+            creation_date = fields.Datetime.from_string(holiday.employee_id.create_date)
+
+            # If employee is created after the period, we cancel the computation
+            if period_end <= creation_date:
```

Comment:
Can't we do that check sooner in the loop ? The sooner we do it, the less useless computation we are doing. 'leaves = holiday.number_per_interval / holiday.employee_id.resource_calendar_id.hours_per_day` might be doing SQL queries to not use the result.